File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/backtrack/P40_CombinationSumII.java
package com.dsa.leetcode.labuladong.backtrack;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * @since 2023/11/19 10:42
 */
public class P40_CombinationSumII {
    List<List<Integer>> ans = new LinkedList<>();
    LinkedList<Integer> path = new LinkedList<>();
    int pathSum = 0;

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtrack(candidates, target, 0);
        return ans;
    }

    private void backtrack(int[] candidates, int target, int start) {
        if (pathSum == target) {
            ans.add(new LinkedList<>(path));
            return;
        }
        if (pathSum > target || start == candidates.length) {
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (i > start && candidates[i - 1] == candidates[i]) {
                continue;
            }
            pathSum += candidates[i];
            path.add(candidates[i]);
            backtrack(candidates, target, i + 1);
            path.removeLast();
            pathSum -= candidates[i];
        }
    }

    public static void main(String[] args) {
        int[] candidates = {10, 1, 2, 7, 6, 1, 5};
        int target = 8;
        P40_CombinationSumII p40CombinationSumII = new P40_CombinationSumII();
        System.out.println(p40CombinationSumII.combinationSum2(candidates, target));
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/backtrack/P40_CombinationSumII.py
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans, path = [], []
        candidates.sort()
        self.backtrack(candidates, target, 0, ans, path)
        return ans

    def backtrack(self, candidates, restTarget, start, ans, path):
        if restTarget == 0:
            ans.append(path[:])
            return
        if restTarget < 0 or start == len(candidates):
            return
        for i in range(start, len(candidates)):
            if i > start and candidates[i - 1] == candidates[i]:
                continue
            path.append(candidates[i])
            self.backtrack(candidates, restTarget - candidates[i], i + 1, ans, path)
            path.pop()


if __name__ == '__main__':
    candidates = [10, 1, 2, 7, 6, 1, 5]
    target = 8
    solu = Solution()
    print(solu.combinationSum2(candidates, target))
